### PR TITLE
Changed configuration validation to allow using a named pipe for port.

### PR DESCRIPTION
**Pull request by hannesne**
Fixes issue #4391

### CODE REVIEW

File: core/server/server.js [hunk 1]
```diff
@@ -28,7 +28,7 @@ const requiredUrl = optionalUrl.required()
 
 const publicConfigSchema = Joi.object({
   bind: {
-    port: Joi.number().port(),
+    port: Joi.optional(),
```
Comment (calebcartwright) [inline core/server/server.js]: Does the below work in the App Service Windows plan env? I'd prefer it if so to keep more explicit validation while supporting that use case

```suggestion
    port: Joi.alternatives().try(
      Joi.number().port(),
      Joi.string(),
    ),
```

Comment (paulmelnikow) [reply core/server/server.js]: We could also try using https://www.npmjs.com/package/unc-path-regex.

Comment (hannesne) [reply core/server/server.js]: That should work fine. I've got a unit test that confirms it as well.

Comment (calebcartwright) [reply core/server/server.js]: @paulmelnikow were you thinking of using `unc-path-regex` as part of a custom `@hapi/joi` validator vs the more lenient `Joi.string()`?

File: core/server/server.spec.js [hunk 2]
```diff
@@ -23,6 +23,14 @@ describe('The server', function() {
     server = undefined
   })
 
+  it('should allow strings for port', async function() {
+    //fixes #4391 - This allows the app to be run using iisnode, which uses a named pipe for the port.
+    const pipeServer = createTestServer({
+      port: '\\\\.\\pipe\\9c137306-7c4d-461e-b7cf-5213a3939ad6',
+    })
+    expect(pipeServer).to.not.undefined
```
Comment (calebcartwright) [inline core/server/server.spec.js]: Not sure if this is a Chai shortcut that I'm just not familiar with or if it should have a `.be` in the chain?

```suggestion
    expect(pipeServer).to.not.be.undefined
```

File: core/server/server.js (hunk 1 above)
Comment (PyvesB) [reply core/server/server.js]: (unresolving this conversation as there were still open questions)

Named pipes happen to be a very specific subset of Windows file paths, the aforementioned `unc-path-regex` will accept them but is much more general. In my opinion, adding a dependency for a single edge-case check on startup would be overkill, especially if that dependency does not do exactly what we want.

Why not write a simple regex that validates against the Windows pipe naming specs? The [Microsoft docs](https://docs.microsoft.com/en-us/windows/win32/ipc/pipe-names) indicates `\\ServerName\pipe\PipeName` and even `\\.\pipe\PipeName` if it's the same server.

File: core/server/server.js [hunk 3]
```diff
@@ -28,7 +28,7 @@ const requiredUrl = optionalUrl.required()
 
 const publicConfigSchema = Joi.object({
   bind: {
-    port: Joi.number().port(),
+    port: Joi.alternatives().try(Joi.number().port(), Joi.string()),
```
Comment (PyvesB) [inline core/server/server.js]:
```suggestion
    port: Joi.alternatives().try(
      Joi.number().port(),
      Joi.string().pattern(/^\\\\\.\\pipe\\.+$/),
    ),
```
Would something like this do the trick?

File: core/server/server.js (hunk 1 above)
Comment (paulmelnikow) [reply core/server/server.js]: If we use a regex we can use `Joi.string().regex()`; we don't need to make an extension. That package is basically one line, though we could add our own regex that is more limited to `\\.\pipe\PipeName`.

File: core/server/server.js (hunk 3 above)
Comment (hannesne) [reply core/server/server.js]: Ok I've tested and applied the change.

**Pull request merged**